source/common/upstream/health_discovery_service.cc

```diff
@@ -102,10 +102,48 @@ void HdsDelegate::handleFailure() {
 // TODO(lilika): Add support for the same endpoint in different clusters/ports
 envoy::service::health::v3::HealthCheckRequestOrEndpointHealthResponse HdsDelegate::sendResponse() {
   envoy::service::health::v3::HealthCheckRequestOrEndpointHealthResponse response;
+
   for (const auto& cluster : hds_clusters_) {
+    // Add cluster health response and set name
+    auto* cluster_health =
+        response.mutable_endpoint_health_response()->add_cluster_endpoints_health();
+    cluster_health->set_cluster_name(cluster->info()->name());
+
+    // Hold all endpoint health information by locality, updated when a new locality
+    // has been seen. When we ingest our configuration and endpoints during the call
+    // of the processMessage() function, all of the endpoints come grouped by cluster
+    // and locality. We continuously appended to the end of the list of endpoints
+    // by locality, meaning each group of endpoints by a particular locality are all
+    // next to each other in the list. If two endpoints next to each other have the same
+    // locality, then we append to the same locality grouping. When the locality
+    // changes from endpoint to endpoint, we start a new grouping by changing this
+    // pointer.
```

Comment:
Could you use `hostsPerLocality()` to make this more explicit?

Reply:
I was having difficulty for a long time getting hostsPerLocality() to work. Previously, there was no notion of the required hosts_per_locality structure being created, and I misunderstood how it was to be created. After you mentioned this I did one last dive into research in how this should be set up, and I now have this working by using hostsPerLocality() by properly setting up the priority set in the constructor. This is reflected in my latest commit. Thanks for the comment!